toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexUpdater.kt

```diff
@@ -0,0 +1,210 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.crates.local
+
+import com.intellij.execution.configuration.EnvironmentVariablesData
+import com.intellij.ide.util.PropertiesComponent
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.runReadAction
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.Task
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.util.registry.Registry
+import com.intellij.util.Alarm
+import com.intellij.util.io.write
+import org.jetbrains.annotations.VisibleForTesting
+import org.rust.cargo.CargoConstants
+import org.rust.cargo.project.model.CargoProjectsService
+import org.rust.cargo.toolchain.BacktraceMode
+import org.rust.cargo.toolchain.RsToolchainBase
+import org.rust.cargo.toolchain.tools.cargo
+import org.rust.openapiext.RsPathManager
+import org.rust.openapiext.RsProcessExecutionException
+import org.rust.openapiext.execute
+import org.rust.stdext.RsResult
+import org.rust.toml.RsTomlBundle
+import java.io.IOException
+import java.nio.file.Path
+import java.util.concurrent.TimeUnit
+import kotlin.io.path.exists
+
+@Service
+class CratesLocalIndexUpdater : Disposable {
+
+    private val alarm: Alarm = Alarm(this)
+
+    private var isUpdating: Boolean = false
+
+    @Synchronized
+    fun updateCratesIoGitIndex() {
```

Comment:
Since `updateCratesIoGitIndex` can only be safely invoked from EDT, let's add an assert and remove `@Synchronized` annotation.

Also, there are several different functions called `updateCratesIoGitIndex` that is a bit confusing.